### PR TITLE
TalentLMS Courses Update v2

### PR DESCRIPTION
**Pull request by dtgreiner**
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
[//]: # (Summarize changes and include links related issue. List any new dependencies)

This updates the Talentlms integration to store the configuration (subdomains) and the courses separately. This is to prevent a bunch of unnecessary API calls as well as decouple the course and configuration to make it easier to call the API for information specific to a course and require specific courses for a user.

Screenshot 2024-10-22 at 9 55 36 AM

The concept of a default course has been added as well as the ability to override the defaults and choose specific courses to be required for a specific user. If training is required of the user, an "Edit Training" link will be available in the user admin list.

Screenshot 2024-10-22 at 10 58 16 AM

Screenshot 2024-10-22 at 10 53 28 AM

_Note: This PR was created due to rebase issues in https://github.com/greenriver/hmis-warehouse/pull/4853._

## Type of change
[//]: # (e.g., Bug fix, New feature, Code clean-up, Dependency update)
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] If adding a new endpoint / exposing data in a new way, I have:
  - [ ] ensured the API can't leak data from other data sources
  - [ ] ensured this does not introduce N+1s
  - [ ] ensured permissions and visibility checks are performed in the right places
- [ ] I have updated the documentation (or not applicable)
- [ ] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

### CODE REVIEW

File: app/controllers/admin/talentlms_controller.rb [hunk 1]
```diff
@@ -7,42 +7,47 @@
...
     def update
-      @config = config_scope.first_or_initialize
       @config.update(config_params)
-      respond_with(@config, location: edit_admin_talentlm_path(@config))
+      respond_with(@config, location: admin_talentlms_path)
+    end
+
+    def destroy
+      @config.courses.each(&:destroy)
+      @config.destroy
```
Comment (eanders) [inline app/controllers/admin/talentlms_controller.rb]: This is fine, but could also be accomplished by adding `dependent: :destroy` to the courses relation on config.
https://guides.rubyonrails.org/v3.2.2/association_basics.html

File: app/controllers/admin/talentlms_courses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module Admin
+  class TalentlmsCoursesController < ApplicationController
+    before_action :require_can_manage_config!
+    before_action :set_course, only: [:update, :edit, :destroy]
+
+    def new
+      config = Talentlms::Config.find(params['config']) if params['config']
```
Comment (eanders) [inline app/controllers/admin/talentlms_courses_controller.rb]:
```suggestion
      config = Talentlms::Config.find(params['config'].to_i) if params['config']
```

Comment (eanders) [reply app/controllers/admin/talentlms_courses_controller.rb]: or maybe config is a string?  If it is, this should get passed through a method that explicitly specifies acceptable values.

File: app/models/talentlms/facade.rb [hunk 3]
```diff
@@ -109,31 +110,34 @@ def create_or_update_local_login(user, lms_account_data)
     #
     # @param user [User] the user
```
Comment (eanders) [inline app/models/talentlms/facade.rb]:
```suggestion
    # @param api [???] the user
```

File: app/models/talentlms/facade.rb [hunk 4]
```diff
@@ -48,43 +49,43 @@ def lms_username(user)
...
+    def lms_email
+      return @user.talent_lms_email if @user.talent_lms_email.present?
 
-      "#{lms_username(user)}@#{ENV['FQDN']}"
+      "#{lms_username}@#{ENV['FQDN']}"
     end
 
     # Login user to TalentLMS
     #
     # @@param user [User] the user
```
Comment (eanders) [inline app/models/talentlms/facade.rb]:
```suggestion
    # @@param api [???] the user
```

File: app/models/talentlms/facade.rb [hunk 5]
```diff
@@ -48,43 +49,43 @@ def lms_username(user)
...
```
Comment (eanders) [inline app/models/talentlms/facade.rb]:
```suggestion
```

File: app/models/talentlms/facade.rb
Comment (eanders) [inline app/models/talentlms/facade.rb:1]: A bunch of methods changed their signature in this file, can you review and update the inline docs to match?  I called out a couple, but it was pretty much anywhere user -> api.

File: app/views/admin/talentlms/edit.haml [hunk 6]
```diff
@@ -10,4 +10,5 @@
   = f.error_notification
   = render 'form', f: f
   .form_actions.mt-4
-    = f.submit 'Save Configuration'
\ No newline at end of file
+    = f.submit 'Save Subdomain'
+    =# link_to 'Delete Subdomain', admin_talentlm_path(config), method: :delete, data: {confirm: 'Are you sure you want to delete this configuration? It will no longer be usable.'}, class: 'btn btn-danger'
```
Comment (eanders) [inline app/views/admin/talentlms/edit.haml]: Should this be commented out?

File: app/views/admin/talentlms_courses/_form.haml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+= f.input :config_id, as: :select_two, label: 'Subdomain', collection: Talentlms::Config.collection_for_form, selected: @course&.config&.id
+= f.input :name
+= f.input :courseid, label: 'Course ID'
+- label = if Rails.env.production? then 'Months' else 'Days' end
+= f.input :months_to_expiration, label: "#{label} Until Required Recertification", hint: 'Leave blank if recertification is not required'
```
Comment (eanders) [inline app/views/admin/talentlms_courses/_form.haml]: Can we dry up this months/days calculation and just ask the Facade class for it?  I believe there is something similar in there, but might as well just make constant.

File: db/warehouse/migrate/20241017183109_update_talent_config_records.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class UpdateTalentConfigRecords < ActiveRecord::Migration[7.0]
+  def change
```
Comment (eanders) [inline db/warehouse/migrate/20241017183109_update_talent_config_records.rb]:
```suggestion
  def up
```

**Review by eanders — CHANGES_REQUESTED** [review]
Thanks @dtgreiner, this is looking pretty good, mostly updates to the inline docs.

**Comment by dtgreiner** [comment]
@reopeining this one, meant to close https://github.com/greenriver/hmis-warehouse/pull/4853

File: app/controllers/admin/talentlms_controller.rb (hunk 1 above)
Comment (dtgreiner) [reply app/controllers/admin/talentlms_controller.rb]: Thanks! Updated this in https://github.com/greenriver/hmis-warehouse/pull/4858/commits/2c8e52bb860a7362c6f607a4f49e98bc646bb3c3

File: app/controllers/admin/talentlms_courses_controller.rb (hunk 2 above)
Comment (dtgreiner) [reply app/controllers/admin/talentlms_courses_controller.rb]: Looks like I missed cleaning up this one from when I was taking a different approach that had the courses being built from within the config. This line is unnecessary now - I'll strip it out in the next commit.

Edit: The [commit](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/e065588b94f6ebb41b04c0f3cfabf7c8bab2415b) to clean this up

File: app/models/talentlms/facade.rb (hunk 3 above)
Comment (dtgreiner) [reply app/models/talentlms/facade.rb]: Facade method signatures cleaned up [here](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/39c3e4ee2d9ce5efe054b34b18dbe31691481e72)

File: app/models/talentlms/facade.rb (hunk 4 above)
Comment (dtgreiner) [reply app/models/talentlms/facade.rb]: Facade method signatures cleaned up [here](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/39c3e4ee2d9ce5efe054b34b18dbe31691481e72)

File: app/models/talentlms/facade.rb
Comment (dtgreiner) [reply app/models/talentlms/facade.rb:1]: Facade method signatures cleaned up [here](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/39c3e4ee2d9ce5efe054b34b18dbe31691481e72)

File: app/models/talentlms/facade.rb (hunk 5 above)
Comment (dtgreiner) [reply app/models/talentlms/facade.rb]: Facade method signatures cleaned up [here](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/39c3e4ee2d9ce5efe054b34b18dbe31691481e72)

File: app/views/admin/talentlms/edit.haml (hunk 6 above)
Comment (dtgreiner) [reply app/views/admin/talentlms/edit.haml]: This comment can be removed, we are deleting configs from the index page.

Cleanup commit: https://github.com/greenriver/hmis-warehouse/pull/4858/commits/fcabe1f421df978bae98d7cf88f84ee6583eb6ba

File: app/views/admin/talentlms_courses/_form.haml (hunk 7 above)
Comment (dtgreiner) [reply app/views/admin/talentlms_courses/_form.haml]: Cleaned this up [here](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/c892c7caa76b3b9252e0fd914a0f04432ab6dfa8)

File: db/warehouse/migrate/20241017183109_update_talent_config_records.rb (hunk 8 above)
Comment (dtgreiner) [reply db/warehouse/migrate/20241017183109_update_talent_config_records.rb]: Updated [here](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/d01ae89a12463349668f22abf1d26865a38ec0b3)

**Comment by dtgreiner** [comment]
@eanders This should be ready for you again. 

I did find an additional bug when going through the cleanup that had to do with not all expired courses being caught as requiring the user to retake the course. - [commit](https://github.com/greenriver/hmis-warehouse/pull/4858/commits/b5ba546fb405ef48508ce64cd02a59820d6f329f)

**Review by eanders — APPROVED** [review]

**Review by eanders — APPROVED** [review]

**Pull request merged**